### PR TITLE
CSCFAIRMETA-507: [ADD] file_name and directory_name parameters for su…

### PR DESCRIPTION
**Pull request by genie9**
…bstring search from directories/{pid}/files

### CODE REVIEW

File: src/metax_api/services/file_service.py [hunk 1]
```diff
@@ -734,19 +748,23 @@ def _get_directory_contents(cls, directory_id, recursive=False, max_depth=1, dep
...
             except Http404:
                 if recursive:
                     return {'directories': []}
                 raise
         else:
             # browsing from ALL files, not cr specific
-            dirs = Directory.objects.filter(parent_directory_id=directory_id).values(*directory_fields)
-
+            dirs = Directory.objects.filter(parent_directory_id=directory_id, directory_name__icontains=directory_name)\
+                .values(*directory_fields)
```
Comment (hannu40k) [inline src/metax_api/services/file_service.py:762]: Had to test how ``directory_name__icontains=''`` is generated into SQL, and this is the result:

`AND UPPER("metax_api_directory"."directory_name"::text) LIKE UPPER('%%'))`

So it looks like if ``directory_name`` is an empty string, the query will have a slight performance cost in every request due to the last two procedure calls. This could be rather easily avoided with something like the following:

    dirs = Directory.objects.filter(parent_directory_id=directory_id).values(*directory_fields)

    if directory_name:
        dirs = dirs.filter(directory_name__icontains=directory_name)

As a side note, the above does not yet execute the query, so even more filters can be added later before executing if necessary.

File: swagger/swagger.yaml [hunk 2]
```diff
@@ -578,22 +578,32 @@ paths:
...
+        - name: file_name
+          in: query
+          description: Substring search from file names. Only matching files are returned.
+          required: false
+          type: string
+        - name: directory_name
+          in: query
+          description: Substring search from directory names. Only matching directories are returned.
+          required: false
+          type: string
```
Comment (hannu40k) [inline swagger/swagger.yaml:590]: Both parameters could include a note about search scope, which I think is the current directory?

File: src/metax_api/services/file_service.py [hunk 3]
```diff
@@ -855,10 +857,16 @@ def _get_directory_contents_for_catalog_record(cls, directory_id, cr_id, file_na
...
+        dirs = Directory.objects.filter(id__in=directory_ids).values(*directory_fields)
+
+        # icontains returns exception on None and with empty string does unnecessary db hits
+        if directory_name:
+            dirs = dirs.filter(directory_name__icontains=directory_name)
+
+        files = None if dirs_only else File.objects.filter(id__in=file_ids).values(*file_fields)
+
+        if file_name:
+            files = files.filter(file_name__icontains=file_name)
```
Comment (hannu40k) [inline src/metax_api/services/file_service.py]: Will this crash if ``dirs_only`` is ``True``, and ``file_name`` is defined, since the last line is then essentially ``None.filter(...)``? Or can this combination exist?

Comment (genie9) [reply src/metax_api/services/file_service.py]: That's a bug spotted and now corrected. Related test added.

**Comment by genie9** [comment]
rebased and ready for merge

File: .gitignore [hunk 4]
```diff
@@ -1,11 +1,14 @@
 *~
 *.pyc
 *.DS_Store
-/docs/*/build/*
+/docs/*build/*
```
Comment (tompulli) [inline .gitignore:4]: Why this? I have my docs build now unignored

File: src/metax_api/migrations/0022_auto_20200527_1639.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
+# Generated by Django 2.2.9 on 2020-05-27 13:39
```
Comment (tompulli) [inline src/metax_api/migrations/0022_auto_20200527_1639.py:1]: Is this file relevant in this ticket?

File: swagger/v2/swagger.yaml [hunk 6]
```diff
@@ -139,6 +139,12 @@ paths:
           description: Filters files by path. Requires query parameter project_identifier.
           required: false
           type: string
+        - name: pagination
```
Comment (tompulli) [inline swagger/v2/swagger.yaml:142]: I guess this was missing from some previous PR and now added here?

File: swagger/v1/swagger.yaml [hunk 7]
```diff
@@ -605,6 +605,16 @@ paths:
...
+        - name: file_name
+          in: query
+          description: Substring search from file names in current directory. With recursive filter also matches files in subdirectories. Only matching files are returned.
+          required: false
+          type: string
+        - name: directory_name
+          in: query
+          description: Substring search from directory names in current directory. With recursive filter also matches subdirectories. Only matching directories are returned.
+          required: false
+          type: string
```
Comment (tompulli) [inline swagger/v1/swagger.yaml:617]: This could be added to v2 also

File: src/metax_api/initialdata/datacatalogs.json [hunk 8]
```diff
@@ -147,6 +147,8 @@
     }
 },
 {
+    "catalog_record_group_edit": "pas_edit",
+    "catalog_record_group_create": "pas_create",
```
Comment (tompulli) [inline src/metax_api/initialdata/datacatalogs.json:151]: Is this relevant to this ticket?

File: .gitignore [hunk 9]
```diff
@@ -1,11 +1,14 @@
...
 /src/__pycache__/*
 /src/static/*
 /src/.coverage
+/swagger/*index.html
```
Comment (tompulli) [inline .gitignore:8]: Is this necessary?

File: src/metax_api/tests/api/rest/v2/views/directories/read.py [hunk 10]
```diff
@@ -992,3 +993,153 @@ def test_read_directory_with_parent_and_pagination(self):
...
+    Test browsing files with queries on file and directory names.
+    """
+
+    def setUp(self):
+        self._use_http_authorization()
+        self._create_test_dirs(5)
+
+    def test_browsing_directory_with_file_name(self):
+
+        response = self.client.get('/rest/directories/24/files?file_name=')
```
Comment (tompulli) [inline src/metax_api/tests/api/rest/v2/views/directories/read.py:1009]: v2 tests should point to v2 api

File: src/metax_api/migrations/0022_auto_20200527_1639.py (hunk 5 above)
Comment (genie9) [reply src/metax_api/migrations/0022_auto_20200527_1639.py:1]: It's no, some how got here after testing another pr.

File: src/metax_api/tests/api/rest/v2/views/directories/read.py (hunk 10 above)
Comment (genie9) [reply src/metax_api/tests/api/rest/v2/views/directories/read.py:1009]: True that!

File: .gitignore (hunk 9 above)
Comment (genie9) [reply .gitignore:8]: gitignore didn't match swagger/index.html and swagger/v2/index.html with swagger/*index.html, so had to add two different lines for those.

File: swagger/v2/swagger.yaml (hunk 6 above)
Comment (genie9) [reply swagger/v2/swagger.yaml:142]: Rebase does sometimes strange things... This branch was rebased so many times that have no idea with this one either why it's here, but decided to keep it so it won't disappear.

File: .gitignore (hunk 4 above)
Comment (genie9) [reply .gitignore:4]: Isn't it suppose to be ignored? I believe repository should have visible build/v1/source and build/v2/source.

File: swagger/v2/swagger.yaml (hunk 6 above)
Comment (genie9) [reply swagger/v2/swagger.yaml:142]: Should be in next PR... will delete this
